### PR TITLE
sandboxes restore on server start

### PR DESCRIPTION
**Pull request by runcom**
Address part of https://github.com/kubernetes-incubator/ocid/issues/17 - sandboxes restore.
Address part of https://github.com/kubernetes-incubator/ocid/issues/31 - pods ids and names

```bash
# start clean...
$ sudo rm -rf /var/lib/ocid/{sandboxes,containers}
$ sudo /usr/local/sbin/runc kill podsanbox1-infra
$ sudo /usr/local/sbin/runc delete podsandbox1-infra

$ sudo ./ocid --runtime /usr/local/sbin/runc --debug &
$ sudo ./ocic pod create --config testdata/sandbox_config.json
$ sudo killall ocid
# podsandbox1 will be in memory again
$ sudo ./ocid --runtime /usr/local/sbin/runc --debug &
# test it by running a container
$ sudo ./ocic container create --pod podsandbox1 --config testdata/container_config.json
# everything should work
# before this patch it was complaining about podsandbox1 not being present
```

@mrunalp PTAL - note that containers restore isn't yet implemented here, possibly a follow up

Signed-off-by: Antonio Murdaca <runcom@redhat.com>

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -24,6 +28,50 @@ type Server struct {
...
+			labels:     m.Labels,
+			containers: make(map[string]*oci.Container),
+		})
+		scontainer, err := oci.NewContainer(m.ContainerName, path, path, m.Labels, sname, false)
+		if err != nil {
+			return err
+		}
+		s.addContainer(scontainer)
+		if err = s.runtime.UpdateStatus(scontainer); err != nil {
+			return err
```
Comment (mrunalp) [inline server/server.go]: Should probably log and continue here.

File: server/runtime.go [hunk 2]
```diff
@@ -177,6 +179,24 @@ func (s *Server) CreatePodSandbox(ctx context.Context, req *pb.CreatePodSandboxR
...
+		ContainerName: containerName,
+		Labels:        labels,
+	}
+
+	b, err := json.Marshal(meta)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: eventually we would track all containers in this pod so on server start
```
Comment (mrunalp) [inline server/runtime.go]: I was thinking that we add the pod id as annotation in the containers so we can link them back. That way we don't need to keep a separate file.

Comment (runcom) [inline server/runtime.go]: I've seen the issue you opened about IDs and yes, sounds good

Comment (runcom) [inline server/runtime.go]: @mrunalp though, we should still save some metadata about the containers as well - like log path, labels and tty - or can all of that come from `config.json`?

Comment (runcom) [inline server/runtime.go]: alright, yeah, everything comes from `config.json` so there won't be the need for the metadata file in containers (probably gonna remove it in sandbox if it isn't needed there as well)

**Comment by mrunalp** [comment]
Yes we will need to store some. Config.json doesn't have everything.

**Comment by runcom** [comment]
> Yes we will need to store some. Config.json doesn't have everything.

it seems `config.json` is sufficient for containers - I'll see if it's the same for sandboxes but we can likely add annotations to `config.json` for almost everything, right?

File: server/server.go (hunk 1 above)
Comment (runcom) [inline server/server.go]: fixed

**Comment by mrunalp** [comment]
@runcom Sounds good. Yes need to audit and see what if anything needs to be stored in metadata.json. Also, we can add anything to annotations.

**Comment by mrunalp** [comment]
We should probably namespace our annotations though like ocid/pod_sandbox_id.

**Comment by runcom** [comment]
Yeah, I will fix the annotation prefix once I do anothor pass at this PR (probably tomorrow morning) - I will probably take care of #31 here as well.

**Comment by runcom** [comment]
@mrunalp took care of having an id and name index for pods - restore for pods still works correctly. I'll address containers ids and names in a follow up to this - this is already too huge

PTAL

**Comment by mrunalp** [comment]
okay reviewing/testing this now.

File: server/sandbox.go [hunk 3]
```diff
@@ -111,7 +121,17 @@ func (s *Server) CreatePodSandbox(ctx context.Context, req *pb.CreatePodSandboxR
 	g.AddBindMount(resolvPath, "/etc/resolv.conf", "ro")
 
 	labels := req.GetConfig().GetLabels()
+	labelsJSON, err := json.Marshal(labels)
+	if err != nil {
+		return nil, err
+	}
+	g.AddAnnotation("oci.labels", string(labelsJSON))
```
Comment (mrunalp) [inline server/sandbox.go]: We would want to do ocid/labels and so on. But this could be a follow on.

Comment (runcom) [inline server/sandbox.go]: Oh crap, I though I had written "ociD" with the "d" - yeah will take care of this one in a follow up

Comment (runcom) [inline server/sandbox.go]: the dot is totally arbitrary - so yeah let's rename it in a follow up

**Comment by mrunalp** [comment]
LGTM

**Comment by runcom** [comment]
*finger crossed* lol

**Pull request merged**